src/core/structures/PDFAcroForm.ts

```diff
@@ -0,0 +1,59 @@
+import {
+  PDFArray,
+  PDFBool,
+  PDFContext,
+  PDFDict,
+  PDFName,
+  PDFNumber,
+  PDFStream,
+  PDFString,
+  XFAResource,
+} from 'src/core';
+import { DictMap } from 'src/core/objects/PDFDict';
+
+class PDFAcroForm extends PDFDict {
```

Comment:
This structure is going to be a bit problematic. Let me explain why:

Currently, the `PDFDict` subclass structures (`PDFCatalog`, `PDFPageLeaf`, and `PDFPageTree`) all have required `/Type` entries. This is important, because it allows us to detect what type of class a given `PDFDict` should be when we are parsing PDF files:

https://github.com/Hopding/pdf-lib/blob/8346da359c1c35089b2d02af07d29228e89482a3/src/core/parser/PDFObjectParser.ts#L194-L204

This means, for example, that the `PDFContext` for a `PDFDocument` will always contain `PDFPageTree` classes, not just `PDFDict` objects with `/Type /Pages`. So we could do this:

```js
const pdfDoc = PDFDocument.load(pdfBytes)
const pageTrees = pdfDoc
  .context
  .enumerateIndirectObjects()
  .find(([, obj]) => obj instanceof PDFPageTree);
```

But, unfortunately, the PDF specification does not define a required `/Type` for AcroForm dictionaries. This makes it very difficult to know whether or not a particular `PDFDict` should be converted to a `PDFAcroForm` when parsing a file. And if we can't create the `PDFAcroForm` instances when parsing, the only other place we can do that is when the `PDFCatalog.AcroForm` method is called. Of course, this would mean that the `PDFAcroForm` structure would simply be a wrapper around its underlying `PDFDict`. And `PDFAcroForm` objects could not be stored in the `PDFContext` object itself.

At best, this violates the convention set by `PDFCatalog`, `PDFPageLeaf`, and `PDFPageTree`. At worst, this makes the behavior of pdf-lib inconsistent and therefore difficult to reason about.

This issue is a bit difficult to explain. I hope I was able to communicate it clearly for you. Please let me know if you don't understand.

I'm not sure what the best way to handle this is. My initial reaction is that the approach pdf-lib currently uses to convert `PDFDict` objects to specialized types when parsing (checking their `/Type` entry) just isn't going to work going forward. And that it would be best to do away with that approach and adhere to a more flexible convention going forward. Of course, I also am cautious about making breaking API/behavior changes. So I'm still weighing the options. If you have any thoughts on the matter, I'd love to hear them!

Reply:
@Hopding I see the issue, but I'm afraid I probably don't know enough about the PDF spec to prescribe a great solution. Writing custom parsing logic for the Catalog seems the best way to go to me, and I think writing structure-level parsing logic would be moving toward the more flexible convention you mentioned. But please let me know if I'm misunderstanding.

Reply:
@cshenks Can you clarify what you mean by the following? I'm not quite sure I follow.
> I think structure-level (as opposed to object-level) parsing for the Catalog would be a step toward this, and would let us cover parsing the AcroForms dictionary separately from ordinary `PDFDict` structures. But please let me know if I'm misunderstanding.

Reply:
@cshenks 

After thinking about this for awhile, the simplest approach I can think of is to avoid extending `PDFDict` and to instead create wrappers for it. This way the parser doesn't need to know anything about the specialized wrapper types. It will just parse into `PDFDict`, `PDFArray`, etc... and leave it at that. Then, when a particular getter method is called, an instance of the specialized wrapper type can be instaniated (backed by the appropriate `PDFDict` instance). And, of course, these wrapper types would not be stored in the `PDFContext`.

I was curious how other PDF libraries handled this problem. (Of course, many PDF libraries only generate documents, so they can avoid this issue entirely). I took a look at the source code for [PDFBox](https://github.com/apache/pdfbox) and discovered that they use the same wrapper approach I described above. 

Here's their version of `PDFDict`: [`COSDictionary`](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/cos/COSDictionary.java). Their parser [generates instances of this class](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/pdfparser/BaseParser.java#L197-L202
), without any sort of specialized type conversion/casting: 

```java
protected COSDictionary parseCOSDictionary() throws IOException
{
    readExpectedChar('<');
    readExpectedChar('<');
    skipSpaces();
    COSDictionary obj = new COSDictionary();
```

They have several `COSDictionary` wrapper classes in the [`pdmodel`](https://github.com/apache/pdfbox/tree/trunk/pdfbox/src/main/java/org/apache/pdfbox/pdmodel) directory. For example:

* [`PDDocumentCatalog`](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/PDDocumentCatalog.java)
* [`PDPage`](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/PDPage.java)
* [`PDAcroForm`](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/interactive/form/PDAcroForm.java)

And they instantiate these wrappers when the corresponding getter methods are invoked:

**[`PDDocumentCatalog`](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/PDDocument.java#L762-L778)**

```java
public PDDocumentCatalog getDocumentCatalog()
{
    if (documentCatalog == null)
    {
        COSDictionary trailer = document.getTrailer();
        COSBase dictionary = trailer.getDictionaryObject(COSName.ROOT);
        if (dictionary instanceof COSDictionary)
        {
            documentCatalog = new PDDocumentCatalog(this, (COSDictionary) dictionary);
        }
        else
        {
            documentCatalog = new PDDocumentCatalog(this);
        }
    }
    return documentCatalog;
}
```

**[`PDPage`](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/PDPageTree.java#L228-L236)**

```java
public PDPage get(int index)
{
    COSDictionary dict = get(index + 1, root, 0);

    sanitizeType(dict);

    ResourceCache resourceCache = document != null ? document.getResourceCache() : null;
    return new PDPage(dict, resourceCache);
}
```

**[`PDAcroForm`](https://github.com/apache/pdfbox/blob/f9b02631102e841668a464c5699b7568cc8087f3/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/PDDocumentCatalog.java#L108-L116)**

```java
public PDAcroForm getAcroForm()
{
    if (cachedAcroForm == null)
    {
        COSDictionary dict = (COSDictionary)root.getDictionaryObject(COSName.ACRO_FORM);
        cachedAcroForm = dict == null ? null : new PDAcroForm(document, dict);
    }
    return cachedAcroForm;
}
```

So considering that we each arrived at a similar solution for the same problem independently, it seems likely to be the best solution, all things considered. Or at least better than the obvious alternatives.

The only complicating factor here for pdf-lib is that there is existing code that relies on the subclassing approach. And while converting this code to the wrapper approach would keep things consistent, it would also introduce a breaking change. So I think it would be best to leave the existing subclassing code as it is (for `PDFCatalog`, `PDFPageLeaf`, and `PDFPageTree`) and just use the wrapper approach for all new models.

Reply:
> @cshenks Can you clarify what you mean by the following? I'm not quite sure I follow.
> 
> > I think structure-level (as opposed to object-level) parsing for the Catalog would be a step toward this, and would let us cover parsing the AcroForms dictionary separately from ordinary `PDFDict` structures. But please let me know if I'm misunderstanding.

I think I was trying to say that if the document were parsed recursively, beginning with the root, it would be possible to parse the AcroForm dictionary based on its relation to the root without having a `/Type` field. This would not work with the existing parsing paradigm, but you could do another pass over the context after parsing to replace `root -> Catalog -> AcroForm` with a `PDFAcroForm` object, if one exists.

Reply:
@Hopding This seems good to me. I'll update `PDFAcroForm` to wrap `PDFDict`